### ISSUE TITLE
Could we make a feature to turn .air into self packaged install without needing to make .swf to do it

### ISSUE THREAD

**Issue by goodnewsjimdotcom** [airsdk/Adobe-Runtime-Support#2461]
Before reporting this issue:

- you should consider whether this would be a question better suited to discussions above ^^ ;
- make sure you have searched for similar feature requests and add your ideas there if one exists.

## Description

Describe your feature request in detail. 

- Will this change the current API? How?
- Who will benefit from this feature?

**Comment by ajwfrost** [comment]
Just to confirm here, the .air file is basically just a zip that includes your SWFs etc, but doesn't include the AIR runtime. So you can then use ADT to turn it into one of:
- native installer package that doesn't include the AIR runtime: upon installation, if the runtime isn't present it would try to download it from Adobe's server which I believe will fail..
- application bundle, which is the local folder structure necessary to run the app (including the AIR runtime), which you can then turn into an installer using separate platform-specific tools

Given we're pushing the option to bundle up the runtimes with the applications, trying to reduce the dependency of the shared AIR runtime, then I would agree that there's a feature missing which is a combination of these, i.e.
- to create a native installer package that also includes the AIR runtime so that it becomes a completely standalone installer that doesn't require any pre-existing software or additional downloads

This is something we've talked about a few times, will see if we can get some bandwidth to do it...

thanks

**Comment by xiaoChi888** [comment]
Great idea! But it may be not so practical.

Only when user first time install AIR application without AIR runtime, the precept would be useful. After that, the precept just increases file size.

However, there's a joke in Chinese market, "Actually you have setup many browsers on your computer". 'Cause many applications in China are based on browser kernel.

Some of my suggestion is to offer an option that packaging an installation program having unified interface (better allow developers custom some icon or image etc).

**Comment by ajwfrost** [comment]
Yes, I agree, it would increase file size and potential duplication of the AIR runtime file on machines (and yes I know what you mean about the number of Chromium DLLs on my computer...). It has a slight advantage in that developers can then fix on a specific version of AIR and avoid problems that may be caused by future updates; this isn't something that should actually be a concern very much, the removal of WebKit was a breaking change but I don't think there would be a need for any further deprecation..

We've looked into this a bit more: what actually happens is:
- if there's no appropriate AIR runtime on the machine, the 'native' installer will go off and download it
- it then runs the downloaded file, passing itself as a command-line argument
- the downloaded file is actually a self-extracting zip which includes the AIR runtime and the runtime and application installer logic, so this will install the AIR runtime and the application one after the other.

We have a slight issue in how the self-extracting archive works, because it's now getting rejected by Apple's code-signing logic due to the way it's packaged. But with a bit of a change to the native installers, I think we can probably get this mechanism working again... and maybe we can also offer an option to create a native installer that includes the runtime directly but in such a way that it's installed as a shared runtime. Then, if the shared runtime is there, it will be used, if not, it will be installed without the need for a download; either way, the installer package may be larger but it would still mean the applications themselves don't have their own individual copies of the AIR runtime libraries..

thanks

**Comment by goodnewsjimdotcom** [comment]
The problem is STEAM demands native installers.

    On Wednesday, March 1, 2023 at 04:09:46 AM EST, Andrew Frost ***@***.***> wrote:  
 
 


Yes, I agree, it would increase file size and potential duplication of the AIR runtime file on machines (and yes I know what you mean about the number of Chromium DLLs on my computer...). It has a slight advantage in that developers can then fix on a specific version of AIR and avoid problems that may be caused by future updates; this isn't something that should actually be a concern very much, the removal of WebKit was a breaking change but I don't think there would be a need for any further deprecation..

We've looked into this a bit more: what actually happens is:
   
   - if there's no appropriate AIR runtime on the machine, the 'native' installer will go off and download it
   - it then runs the downloaded file, passing itself as a command-line argument
   - the downloaded file is actually a self-extracting zip which includes the AIR runtime and the runtime and application installer logic, so this will install the AIR runtime and the application one after the other.

We have a slight issue in how the self-extracting archive works, because it's now getting rejected by Apple's code-signing logic due to the way it's packaged. But with a bit of a change to the native installers, I think we can probably get this mechanism working again... and maybe we can also offer an option to create a native installer that includes the runtime directly but in such a way that it's installed as a shared runtime. Then, if the shared runtime is there, it will be used, if not, it will be installed without the need for a download; either way, the installer package may be larger but it would still mean the applications themselves don't have their own individual copies of the AIR runtime libraries..

thanks
